### PR TITLE
Replace react dom's testing library with ReactTestingLibrary

### PR DESCRIPTION
**Pull request by davesnx**
Depends on #846 

- Vendors ReactTestingLibrary (and does some ad-hoc changes) because React 19 [deprecates react-dom test utils](https://react.dev/warnings/react-dom-test-utils#rest-of-reactdomtestutils-apis).
- Creates a `demo` folder to ensure React works manually, loads, etc.
- Migrates all jest tests to use RTL
- Currently uses React.Uncurried for useState and useReducer waiting for https://github.com/reasonml/reason-react/pull/860 and to have a better solution from https://github.com/facebook/react/issues/31578

### CODE REVIEW

File: test/melange-testing-library/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+### vendored melange-testing-library
```
Comment (jchavarri) [inline test/melange-testing-library/README.md:1]: I read the PR description but it explains what's being done but not why.

What was the issue with using react test renderer? It's a much lighter solution than react testing library and it should catch any runtime issues just fine.

File: test/melange-testing-library/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+
+This directory contains a vendored version of the melange-testing-library.
+
+The original repository is https://github.com/melange-community/melange-testing-library
```
Comment (jchavarri) [inline test/melange-testing-library/README.md:5]: The PR description mentions

> Vendors ReactTestingLibrary (and does some ad-hoc changes)

are the ad-hoc changes valuable for other users of the bindings? in that case, will they be upstreamed?

**Review by jchavarri — COMMENTED** [review]
I'm afraid I don't know the reason why switching to another (heavier) testing library is needed.

File: test/melange-testing-library/README.md (hunk 2 above)
Comment (davesnx) [reply test/melange-testing-library/README.md:5]: Not sure, melange-testing-library contains a few binding designs focused on previous versions of BuckleScript where options aren't translated to undefined, or records aren't objects.

I would keep the changes here and evolve this bindings as a playground for the melange-testing-library

File: test/melange-testing-library/README.md (hunk 1 above)
Comment (jchavarri) [reply test/melange-testing-library/README.md:1]: ftr: https://react.dev/warnings/react-dom-test-utils#rest-of-reactdomtestutils-apis

File: Makefile [hunk 3]
```diff
@@ -54,11 +59,19 @@ format-check: ## Checks if format is correct
 .PHONY: install
 install: ## Update the package dependencies when new deps are added to dune-project
 	@opam install . --deps-only --with-test
-	@npm install
+	@npm install --force
```
Comment (jchavarri) [inline Makefile:62]: Why was `force` needed?

File: demo/dune [hunk 4]
```diff
@@ -0,0 +1,9 @@
+(melange.emit
+ (target demo)
+ (alias melange-app)
+ (module_systems
+  (es6 mjs))
+ (libraries reason-react jest melange.belt melange.dom)
```
Comment (jchavarri) [inline demo/dune:6]: Why is `jest` needed for the demo?

**Review by jchavarri — APPROVED** [review]

File: demo/dune (hunk 4 above)
Comment (davesnx) [reply demo/dune:6]: Not needed indeed

File: Makefile (hunk 3 above)
Comment (davesnx) [reply Makefile:62]: Because testinglib and other deps depend on react 18, so npm complains. It is a decent workaround to ensure the installation succeeds

Comment (davesnx) [reply Makefile:62]: Will write a comment

File: test/melange-testing-library/README.md (hunk 2 above)
Comment (anmonteiro) [reply test/melange-testing-library/README.md:5]: works, but we should upstream as they get better!

File: Makefile (hunk 3 above)
Comment (anmonteiro) [reply Makefile:65]: would we still needed `--force` if you commit the lockfile from manual install and run `npm ci` here?

**Review by anmonteiro — APPROVED** [review]

**Pull request merged**